File: AdministracionDeTienda/src/gestorAplicacion/clasesBase/Bodega.java
package gestorAplicacion.clasesBase;

import java.util.ArrayList;

import gestorAplicacion.clasesBase.Tienda;


public class Bodega implements Inventariar,java.io.Serializable{

	
	/**
	 * 
	 */
	private final int stopBodega;
	private static final long serialVersionUID = 1L;
	private ArrayList<Producto> productosEnBodega = new ArrayList<Producto>();
	private static String resumenPedido;
	private static ArrayList<Transferencia> pagos=new ArrayList<Transferencia>();
	
	public Bodega(ArrayList<Producto> productos) {
		this.stopBodega = productos.size();
		productosEnBodega=productos;
	}
	public Bodega( int max) {
		stopBodega = max;
	}
	
	public int calcularNumeroProductos() {		
		return productosEnBodega.size();
	}
	
	public void agregarProductos(Producto producto) {
		this.productosEnBodega.add(producto);
	}
	
	public int calcularCamisas() {
		return Inventariar.calcularCamisas(productosEnBodega);
	}
	
	public int calcularAbrigos() {
		return Inventariar.calcularAbrigos(productosEnBodega);
	}
	
	public int calcularPantalon() {
			return Inventariar.calcularPantalon(productosEnBodega);
		
	}
	
	public int calcularCamisas(ArrayList<Producto> pedido) {
		int numeroCamisetas = 0;
		for (int i = 0; i < pedido.size(); i++) {
			if( pedido.get(i).getNombre().equals("CAMISA")) {
				numeroCamisetas ++;
			}
		}
		return numeroCamisetas;
	}
	
	
	public int calcularPantalon(ArrayList<Producto> pedido) {
		int numeroPantalones = 0;
		for (int i = 0; i < pedido.size(); i++) {
			if(pedido.get(i).getNombre().equals("PANTALON")) {
				numeroPantalones ++;
			}
		}
		return numeroPantalones;
	}
	
	
	public int calcularAbrigos(ArrayList<Producto> pedido) {
		int numeroAbrigo = 0;
		for (int i = 0; i < pedido.size(); i++) {
			if(pedido.get(i).getNombre().equals("ABRIGO")) {
				numeroAbrigo ++;
			}
		}
		return numeroAbrigo;
	}
	
	
	
	
	public ArrayList<Producto> getProductosEnBodega() {
		return productosEnBodega;
	}


	public void setProductosEnBodega(ArrayList<Producto> productosEnBodega) {
		this.productosEnBodega = productosEnBodega;
	}


	public int getMinCamisas() {
		return minCamisas;
	}


	public int getMinAbrigo() {
		return minAbrigo;
	}


	public int getMinPantalon() {
		return minPantalon;
	}
public ArrayList<Producto> buscarProducto(ArrayList<Producto>productos) {
		
		ArrayList<Producto> listaAuxiliar=new ArrayList<Producto>(productosEnBodega);
		ArrayList<Producto>productosBuscados=new ArrayList<Producto>();
		for(Producto p: productos) {
			for (Producto p1:listaAuxiliar) {
				if(p.equals(p1)){
					productosBuscados.add(p1);
					listaAuxiliar.remove(p1);
					break;
				}
			}
		}
		return productosBuscados;
		
	}



public boolean evaluarDisponibilidad(ArrayList<Producto>productosRetirar) {
	int contadorAbrigo=Inventariar.calcularAbrigos(productosRetirar);
	int contadorPantalon=Inventariar.calcularPantalon(productosRetirar);
	int contadorCamisas=Inventariar.calcularCamisas(productosRetirar);
	

	
	if(this.calcularCamisas()-contadorCamisas>=minCamisas && 
	this.calcularAbrigos()-contadorAbrigo>=minAbrigo && 
	this.calcularPantalon()-contadorPantalon>=minPantalon) {
		
		this.retirarProductos(productosRetirar);
		return true;
		
	}else {
		return false;
	}
	}
		
	

	private void retirarProductos(ArrayList<Producto> productos) {	

			ArrayList<String>borrarProducto=new ArrayList<String>();
			ArrayList<Producto>productosBorrar=new ArrayList<Producto>();
			for(Producto p1:productos) {
				borrarProducto.add(p1.getNombre());
			}
			for(Producto p: productosEnBodega) {
				
				if(borrarProducto.indexOf(p.getNombre())!=-1){
					productosBorrar.add(p);
					borrarProducto.remove(p.getNombre());
				}
			}
			
			productosEnBodega.removeAll(productosBorrar);
		
		}
		
	
	
	public int getStopBodega() {return stopBodega;}
	
	
	public String toString() {
		
		return "Camisetas = " + calcularCamisas(getProductosEnBodega())+ "\nPantalones = " + calcularPantalon(getProductosEnBodega()) + "\nAbrigos = " +calcularAbrigos(getProductosEnBodega());
	}
	//Relacionado con ENVIO
	public enum SETS{
		ONLY,TU,COMPLETO;	
	}
	//PRIMERA INTERACCION-ENVIO
	public static ArrayList<Producto> realizarPedido(ArrayList<Tienda> tiendas,SETS set,int cantidad){
		ArrayList<Producto>necesarios=Producto.seleccionarProductos(set,cantidad);
		ArrayList<Producto> pedido=new ArrayList<Producto>(); 
		for(int i=0;i<tiendas.size();i++) {
			if(necesarios.isEmpty()) {
				break;
				
			}else {
				Bodega bodega=tiendas.get(i).getBodega();
				ArrayList<Producto> productos=bodega.pedirProductos(necesarios);
				pedido.addAll(productos);
				gestionarPago(tiendas.get(i),productos);
				
			}	
			
		}
		crearResumen(necesarios,pedido);
		
		
		return pedido;
	}
	
	private static void crearResumen(ArrayList<Producto> necesarios, ArrayList<Producto> pedido) {
		if(necesarios.size()==0) {
			setResumenPedido("El pedido se ha completado exitosamente");
		}else {
			setResumenPedido("Por falta de disponibilidad de productos en las tiendas el pedido solo ha podido completar "+pedido.size()+" de los "+(((int)pedido.size())+((int)necesarios.size()))+" necesarios.");
		}
		
	}
	public static void gestionarPago(Tienda tienda, ArrayList<Producto> productos) {
		double cantidad=0;
		for(int i=0;i<productos.size();i++) {
			cantidad+=productos.get(i).getCosto();
			
		}
		Transferencia pago=new Transferencia(Tienda.getCuentaTienda(),cantidad);
		pago.setDetalle("Id Tienda:"+tienda.getId()+" VALOR: $"+cantidad);
		getPagos().add(pago);
		
		
	}

//Metodo que se llama desde la primera interaccion
	public ArrayList<Producto> pedirProductos(ArrayList<Producto> productos){
		ArrayList<Producto> pedido=new ArrayList<Producto>();
		ArrayList<Producto> completado=new ArrayList<Producto>();
		for(int i=0;i<productos.size();i++) {
			for(int e=0;e<productosEnBodega.size();e++) {
				if(productos.get(i).getTipo()==productosEnBodega.get(e).getTipo()) {
					pedido.add(productosEnBodega.get(e));
					completado.add(productos.get(i));
					productosEnBodega.remove(e);
					break;
				}
			}
			
		}
		productos.removeAll(completado);
		return pedido;
	}

	public void abastecerBodega(ControlCalidad c, ArrayList<Producto> lista) {
        if ((productosEnBodega.isEmpty() || productosEnBodega == null ) && ControlCalidad.listaSinLista(c.getCompra().getCompraLlego(), c.getProductosDefectuosos()) != null) {
            this.setProductos(ControlCalidad.listaSinLista(c.getCompra().getCompraLlego(), c.getProductosDefectuosos()));
            if (lista != null) {
                this.abastecerBodega(lista);
            }
        }  else if (lista != null){
            this.abastecerBodega(lista);
        }

    }  

    public void abastecerBodega(ArrayList<Producto> lista) {
        for (Producto producto : lista) {
            this.addProducto(producto);
        }

    }

	
    public void addProducto(Producto producto){
        productosEnBodega.add(producto);
    }

	public void setProductos(ArrayList<Producto> productos){
        this.productosEnBodega = productos;
    }

	public static String getResumenPedido() {
		return resumenPedido;
	}

	public static void setResumenPedido(String resumenPedido) {
		Bodega.resumenPedido = resumenPedido;
	}

	public static ArrayList<Transferencia> getPagos() {
		return pagos;
	}

	public static void setPagos(ArrayList<Transferencia> pagos) {
		Bodega.pagos = pagos;
	}
	

}

File: Python/src/gestorAplicacion/clasesBase/Bodega.py
from typing import List
from Producto import Producto
from Inventariar import Inventariar
from Transferencia import Transferencia
from ControlCalidad import ControlCalidad
from Tienda import Tienda

class Bodega(Inventariar):
    pagos = []
    resumenPedido = ""


    def __init__(self, productos,stop):
        self.stopBodega = stop
        self.productosEnBodega = productos

    def calcularNumeroProductos(self):
        return len(self.productosEnBodega)

    def agregarProductos(self, producto):
        self.productosEnBodega.append(producto)

    def calcularCamisas(self):
        return Inventariar.calcularCamisas(self.productosEnBodega)

    def calcularAbrigos(self):
        return Inventariar.calcularAbrigos(self.productosEnBodega)

    def calcularPantalon(self):
        return Inventariar.calcularPantalon(self.productosEnBodega)

    def calcularCamisas(self, pedido):
        numeroCamisetas = 0
        for producto in pedido:
            if producto.getNombre() == Producto.Tipo.CAMISA:
                numeroCamisetas += 1
        return numeroCamisetas

    def calcularPantalon(self, pedido):
        numeroPantalones = 0
        for producto in pedido:
            if producto.getNombre() == Producto.Tipo.PANTALON:
                numeroPantalones += 1
        return numeroPantalones

    def calcularAbrigos(self, pedido):
        numeroAbrigo = 0
        for producto in pedido:

            if producto.getNombre() == Producto.Tipo.ABRIGO:

                numeroAbrigo += 1
        return numeroAbrigo

    def getProductosEnBodega(self):
        return self.productosEnBodega

    def setProductosEnBodega(self, productosEnBodega):
        self.productosEnBodega = productosEnBodega

    def getMinCamisas(self):
        return Inventariar.minCamisas

    def getMinAbrigo(self):
        return Inventariar.minAbrigo

    def getMinPantalon(self):
        return Inventariar.minPantalon

    def buscarProducto(self, productos):
        listaAuxiliar = self.getProductosEnBodega().copy()
        productosBuscados = []
        for producto in productos:
            for p in listaAuxiliar:
                if producto == p:
                    productosBuscados.append(p)
                    listaAuxiliar.remove(p)
                    break
        return productosBuscados

    def evaluarDisponibilidad(self, productosRetirar):
        contadorAbrigo = self.calcularAbrigos(productosRetirar)
        contadorPantalon = self.calcularPantalon(productosRetirar)
        contadorCamisas = self.calcularCamisas(productosRetirar)

        if (Inventariar.calcularCamisas(self.productosEnBodega) - contadorCamisas >= self.getMinCamisas() and
            Inventariar.calcularAbrigos(self.productosEnBodega) - contadorAbrigo >= self.getMinAbrigo() and
            Inventariar.calcularPantalon(self.productosEnBodega) - contadorPantalon >= self.getMinPantalon()):
            self.retirarProductos(productosRetirar)
            return True
        else:
            return False

    def retirarProductos(self, productos):
        borrarProducto = [p.getNombre() for p in productos]
        productosBorrar = []
        for p in self.productosEnBodega:
            if p.getNombre() in borrarProducto:
                productosBorrar.append(p)
                borrarProducto.remove(p.getNombre())
        for p in productosBorrar:
            self.productosEnBodega.remove(p)

    def getStopBodega(self):
        return self.stopBodega

    def __str__(self):
        return f"Camisetas = {self.calcularCamisas(self.getProductosEnBodega())}\nPantalones = {self.calcularPantalon(self.getProductosEnBodega())}\nAbrigos = {self.calcularAbrigos(self.getProductosEnBodega())}"

    @staticmethod
    def realizar_pedido(tiendas, set_tipo, cantidad):

        necesarios = Producto.seleccionarProductos(set_tipo, cantidad)
        pedido = []
        for tienda in tiendas:
            if not necesarios:
                break
            else:
                bodega = tienda.getBodega()
                productos = bodega.pedir_productos(necesarios)
                pedido.extend(productos)
                Bodega.gestionar_pago(tienda, productos)
        Bodega.crear_resumen(necesarios, pedido)
        return pedido

    @staticmethod
    def crear_resumen(necesarios, pedido):
        if not necesarios:
            Bodega.resumenPedido = "El pedido se ha completado exitosamente"
        else:
            completados = len(pedido)
            total = completados + len(necesarios)
            Bodega.resumenPedido = f"Por falta de disponibilidad de productos en las tiendas el pedido solo ha podido completar {completados} de los {total} necesarios."

    @staticmethod
    def gestionar_pago(tienda, productos):
        cantidad=0
        for p in productos:
            cantidad+=p.getCosto()

        pago = Transferencia(Tienda.cuentaTienda,Tienda.cuentaTienda.getEntidad(),cantidad)
        pago.setDestinatario(Tienda.cuentaTienda)
        tienda.presupuesto+=cantidad
        Bodega.pagos.append(pago)

    def pedir_productos(self, productos):
        pedido = []
        completado = []
        for producto in productos:
            for i in range(len(self.productosEnBodega)):
                if producto.get_tipo() == self.productosEnBodega[i].get_tipo():
                    pedido.append(self.productosEnBodega[i])
                    completado.append(producto)
                    del self.productosEnBodega[i]
                    break

        for producto in completado:
            productos.remove(producto)

        return pedido

    def abastecerBodega(self, c, lista):
        if (len(self.productosEnBodega) == 0 or self.productosEnBodega is None) and ControlCalidad.listaSinLista(c.getCompra().getCompraLlego(), c.getProductosDefectuosos()) is not None:
            self.setProductos(ControlCalidad.listaSinLista(c.getCompra().getCompraLlego(), c.getProductosDefectuosos()))
            if lista is not None:
                self.abastecerBodega2(lista)
        elif lista is not None:
            self.abastecerBodega2(lista)

    def abastecerBodega2(self, lista):
        for producto in lista:
            self.addProducto(producto)

    def addProducto(self, producto):
        self.productosEnBodega.append(producto)

    def setProductos(self, productos):
        self.productosEnBodega = productos

    @staticmethod
    def getResumenPedido():
        return Bodega.resumenPedido


    def setResumenPedido(resumenPedido):
        Bodega.resumenPedido = resumenPedido


    def getPagos(self):
        return self.pagos

    def setPagos(self,pagos):
        self.pagos = pagos

    # def __str__(self):
    #     return f"Camisas: {self.calcularCamisas()} \nPantalones: {self.calcularPantalon()}"
    #
